sapi/cli/php_cli_server.c

```diff
@@ -1622,56 +1619,45 @@ static int php_cli_server_client_read_request_on_fragment(php_http_parser *parse
-	zend_string *orig_header_name = zend_string_init(client->current_header_name, client->current_header_name_len, 1);
-	zend_string *lc_header_name = zend_string_alloc(client->current_header_name_len, 1);
-	zend_str_tolower_copy(ZSTR_VAL(lc_header_name), client->current_header_name, client->current_header_name_len);
-	GC_MAKE_PERSISTENT_LOCAL(orig_header_name);
-	GC_MAKE_PERSISTENT_LOCAL(lc_header_name);
-	zend_hash_add_ptr(&client->request.headers, lc_header_name, client->current_header_value);
-	zend_hash_add_ptr(&client->request.headers_original_case, orig_header_name, client->current_header_value);
-	zend_string_release_ex(lc_header_name, 1);
-	zend_string_release_ex(orig_header_name, 1);
-
-	if (client->current_header_name_allocated) {
-		pefree(client->current_header_name, 1);
-		client->current_header_name_allocated = 0;
-	}
+	zend_string *perm_header_name = zend_string_dup(client->current_header_name, /* persistent */ true);
+	zend_string *lc_header_name = zend_string_tolower_ex(client->current_header_name, /* persistent */ true);
+
+	/* Add the wrapped zend_string to the HashTable */
+	zend_hash_add(&client->request.headers, lc_header_name, &tmp);
+	zend_hash_add(&client->request.headers_original_case, perm_header_name, &tmp);
+
+	zend_string_release_ex(client->current_header_name, /* persistent */ false);
+	zend_string_release_ex(lc_header_name, /* persistent */ true);
+
 	client->current_header_name = NULL;
-	client->current_header_name_len = 0;
 	client->current_header_value = NULL;
-	client->current_header_value_len = 0;
 }
 
 static int php_cli_server_client_read_request_on_header_field(php_http_parser *parser, const char *at, size_t length)
 {
 	php_cli_server_client *client = parser->data;
 	switch (client->last_header_element) {
-	case HEADER_VALUE:
-		php_cli_server_client_save_header(client);
-		ZEND_FALLTHROUGH;
-	case HEADER_NONE:
-		client->current_header_name = (char *)at;
-		client->current_header_name_len = length;
-		break;
-	case HEADER_FIELD:
-		if (client->current_header_name_allocated) {
-			size_t new_length = client->current_header_name_len + length;
-			client->current_header_name = perealloc(client->current_header_name, new_length + 1, 1);
-			memcpy(client->current_header_name + client->current_header_name_len, at, length);
-			client->current_header_name[new_length] = '\0';
-			client->current_header_name_len = new_length;
-		} else {
-			size_t new_length = client->current_header_name_len + length;
-			char* field = pemalloc(new_length + 1, 1);
-			memcpy(field, client->current_header_name, client->current_header_name_len);
-			memcpy(field + client->current_header_name_len, at, length);
-			field[new_length] = '\0';
+		case HEADER_VALUE:
+			/* Save previous header before creating new one */
+			php_cli_server_client_save_header(client);
+			ZEND_FALLTHROUGH;
+		case HEADER_NONE:
+			/* Create new header field */
+			client->current_header_name = zend_string_init(at, length, /* persistent */ false);
```

Comment:
This one needs to be persistent too

Reply:
otherwise it's freed too early in this case:

```
$fd = stream_socket_client("tcp://localhost:8082");
fwrite($fd, "GET /index.php HTTP/1.0\r\nHost: hello");

$fd2 = stream_socket_client("tcp://localhost:8082");
fwrite($fd2, "GET /index.php HTTP/1.0\r\nConnection: close\r\n\r\n");

stream_copy_to_stream($fd2, STDOUT);

fwrite($fd, "\r\n\r\n");
stream_copy_to_stream($fd, STDOUT);
```

Reply:
I didn't manage to reproduce this locally. Hopefully I fixed the issue.